### ISSUE TITLE
ILU2 folder lacks bcg.f, future considerations

### ISSUE THREAD

**Issue by kirill-terekhov** [INMOST-DEV/INMOST#11]
Please put bcg.f from ANI3D into ILU2 folder.

After I realised that ANI3D solver was barely useful due to constant "lendwork too small or partlur too big" problem, the main purpose of supporting it was to test performance of INNER_ILU2 against initial fortran realisation. At the time I implemented INNER_ILU2 I haven't noticed any significant performance difference between the two. I suggest someone retest that this is still true (one should set overlap 0, because there is some redundant work and ensure that LFILL is undefined since it changes the algorithm) and then we can drop the support of ANI3D ILU2 solver. I cannot currently retest it since I don't have fortran on my windows machine.

**Comment by kirill-terekhov** [comment]
Also for retest purposes one should uncomment `#define KSOLVER BCGS_solver` and comment `#define KSOLVER BCGSL_solver` so that iterative methods will match.

**Comment by a-a-danilov** [comment]
Added BCG files from Ani3D-3.0. CMakeLists.txt needs further adjustments to include BLAS library.
Tried both INNER_ILU2 and ANI solvers on the same matrices using `examples/MatSolve`.
LFILL was not declared, the following change was made, as suggest in previous comment:
```
diff --git a/solver.cpp b/solver.cpp
index d2d7184..bdb1f8a 100644
--- a/solver.cpp
+++ b/solver.cpp
@@ -32,8 +32,8 @@
 #include "Teuchos_XMLParameterListHelpers.hpp"
 #endif
 
-#define KSOLVER BCGSL_solver
-//#define KSOLVER BCGS_solver
+//#define KSOLVER BCGSL_solver
+#define KSOLVER BCGS_solver
 
 
 namespace INMOST
```
For test purposes I used Intel compilers with MKL as BLAS library source.
CMake options: `cmake -DCMAKE_BUILD_TYPE=Release -DTEST_FORTRAN_ANI3D=ON -DCMAKE_C_COMPILER=icc -DCMAKE_CXX_COMPILER=icpc -DCMAKE_Fortran_COMPILER=ifort -DCOMPILE_EXAMPLES=ON -DCMAKE_EXE_LINKER_FLAGS="-mkl -lifcore" ../inmost-github/`

Tried three random matrices from SPARSKIT MM collection: fidap/fidapm11, tokamak/utm5940, drivcav/e40r2000. Note that 0 correspond to INNER_ILU2, and 7 to ANI, only serial runs were performed:
```
$ ./MatSolve 0 fidapm11{,_rhs1}.mtx
load matrix: 0.636319
load vector: 0.0166202
preconditioner: 6.91379
solver: 0.2266772.35191e-06 | 1e-05
||Ax-b|| 2.35191e-06 ||b|| 0.390066 ||Ax-b||/||b|| 6.02951e-06
norms: 0.00998807
1 processors solved in 7.79762 secs with 5 iterations to 2.35191e-06 norm matrix "fidapm11.mtx" vector "fidapm11_rhs1.mtx" true residual ||Ax-b||/||b|| 6.02951e-06
reason: converged due to absolute tolerance

$ ./MatSolve 7 fidapm11{,_rhs1}.mtx
load matrix: 0.67542
load vector: 0.0163012
 DLS =    111837342.409568       DLR =    742304.876770460     
 AFNORM =    149.311754393282     
 LFNORM =    8885.23844988327       UFNORM =    1493.56025368634     
preconditioner: 4.47044
ILU: iters:   5, residual=0.195E-05 (|SOL|= 0.3141E+03, -0.1525E+01 < SOL <  0.7790E+01)
solver: 0.121492
||Ax-b|| 1.94846e-06 ||b|| 0.390066 ||Ax-b||/||b|| 4.9952e-06
norms: 0.010236
1 processors solved in 5.29191 secs with 5 iterations to 1.94846e-06 norm matrix "fidapm11.mtx" vector "fidapm11_rhs1.mtx" true residual ||Ax-b||/||b|| 4.9952e-06
reason: Unspecified for ANI
```
Nearly the same result.

```
$ ./MatSolve 0 utm5940{,_rhs1}.mtx
load matrix: 0.085674
load vector: 0.00514293
preconditioner: 0.412354
solver: 0.67027 4.43039e-06 | 1e-05
||Ax-b|| 4.43039e-06 ||b|| 0.0787991 ||Ax-b||/||b|| 5.62239e-05
norms: 0.0010221
1 processors solved in 1.17395 secs with 128 iterations to 4.43039e-06 norm matrix "utm5940.mtx" vector "utm5940_rhs1.mtx" true residual ||Ax-b||/||b|| 5.62239e-05
reason: converged due to absolute tolerance

$ ./MatSolve 7 utm5940{,_rhs1}.mtx
load matrix: 0.085223
load vector: 0.00550199
 DLS =    46366.9536502970       DLR =    21752.0135254006     
 AFNORM =    77.0713954719912     
 LFNORM =    5102548.64470261       UFNORM =    36863459.7274392     
preconditioner: 0.329026
ILU: iters: 212, residual=0.378E-06 (|SOL|= 0.3644E+06, -0.4066E+04 < SOL <  0.3480E+05)
solver: 0.558265
||Ax-b|| 3.77917e-07 ||b|| 0.0787991 ||Ax-b||/||b|| 4.79596e-06
norms: 0.00112796
1 processors solved in 0.979353 secs with 212 iterations to 3.77886e-07 norm matrix "utm5940.mtx" vector "utm5940_rhs1.mtx" true residual ||Ax-b||/||b|| 4.79596e-06
reason: Unspecified for ANI
```
ANI solver need twice as much iterations as INNER_ILU2.

```
$ ./MatSolve 0 e40r2000{,_rhs1}.mtx
load matrix: 0.589706
load vector: 0.0148902
preconditioner: 4.44604
solver: 4.62712 9.96167e-06 | 1e-05
||Ax-b|| 9.96167e-06 ||b|| 1.81648 ||Ax-b||/||b|| 5.48406e-06
norms: 0.00929189
1 processors solved in 9.68128 secs with 114 iterations to 9.96167e-06 norm matrix "e40r2000.mtx" vector "e40r2000_rhs1.mtx" true residual ||Ax-b||/||b|| 5.48406e-06
reason: converged due to absolute tolerance

$ ./MatSolve 7 e40r2000{,_rhs1}.mtx
load matrix: 0.593357
load vector: 0.015132
 DLS =    8843569.98563548       DLR =    571941177.374187     
 AFNORM =    131.457217375084     
 LFNORM =    247819.342601400       UFNORM =    8116782.06378373     
preconditioner: 3.05262
ILU: iters:  35, residual=0.321E-05 (|SOL|= 0.5607E+03, -0.2514E+02 < SOL <  0.7768E+02)
solver: 0.739327
||Ax-b|| 3.21058e-06 ||b|| 1.81648 ||Ax-b||/||b|| 1.76748e-06
norms: 0.0095799
1 processors solved in 4.40793 secs with 35 iterations to 3.21058e-06 norm matrix "e40r2000.mtx" vector "e40r2000_rhs1.mtx" true residual ||Ax-b||/||b|| 1.76748e-06
reason: Unspecified for ANI
```
ANI needs three times less iterations and is twice faster.

The conclusions are not clear to me. May be you can suggest some more testing scenarios?

**Comment by kirill-terekhov** [comment]
Seems there were other pitfalls:

Due to lack of support, ANI solver have hardcoded tau = 1e-3, tau2 = 1.e-6 and gets tolerance as 1.e-5 * ||b-Ax||. It also performs 5 iterations to converge matrix row/column norms to 1.

On the other side INNER_ILU2 parameters are prescribed from MatSolve example: tau = 1.e-3, tau2 = 1.e-5, and it performs 8 iterations to converge row/column norms. By default it takes absolute tolerance 1.e-5 and relative tolerance 1.0e-12.

MKL may boost bcg performance significantly, since it is vectorized and some version of library is parallelized as well. I remember using built-in ANI3D blas.

**Comment by a-a-danilov** [comment]
So, I retested with INNER_ILU2 parameters matching hardcoded in ANI solver.
Specific changes were made:
```
diff --git a/examples/MatSolve/main.cpp b/examples/MatSolve/main.cpp
index dbb18a4..9303405 100644
--- a/examples/MatSolve/main.cpp
+++ b/examples/MatSolve/main.cpp
@@ -72,14 +72,17 @@ int main(int argc, char ** argv)
                        Solver s(type); // Declare the linear solver by specified type
 
                        s.SetParameterEnum("gmres_substeps",3);
+                       s.SetParameterEnum("schwartz_overlap",0);
                        
                        s.SetParameterEnum("reorder_nonzeros",0);
-                       s.SetParameterEnum("rescale_iterations",8);
+                       s.SetParameterEnum("rescale_iterations",5);
                        s.SetParameterEnum("adapt_ddpq_tolerance",0);
                        
                        s.SetParameterReal("drop_tolerance",0.001);
-                       s.SetParameterReal("reuse_tolerance",0.00001);
+                       s.SetParameterReal("reuse_tolerance",0.000001);
                        s.SetParameterReal("ddpq_tolerance",0.7);
+                       s.SetParameterReal("absolute_tolerance",0);
+                       s.SetParameterReal("relative_tolerance",1e-5);
                        
 
                        t = Timer();
diff --git a/solver.cpp b/solver.cpp
index d2d7184..bdb1f8a 100644
--- a/solver.cpp
+++ b/solver.cpp
@@ -32,8 +32,8 @@
 #include "Teuchos_XMLParameterListHelpers.hpp"
 #endif
 
-#define KSOLVER BCGSL_solver
-//#define KSOLVER BCGS_solver
+//#define KSOLVER BCGSL_solver
+#define KSOLVER BCGS_solver
 
 
 namespace INMOST
```

Compiled with `CMAKE_BUILD_TYPE=Release` with gcc-4.8.3 compilers and reference implementation of BLAS. Tested the same matrices in the same order. The results look now nearly the same.
```
$ ./MatSolve 0 fidapm11{,_rhs1}.mtx
load matrix: 0.939934
load vector: 0.024579
preconditioner: 7.84237
solver: 0.1781362.22133e-06 | 0
||Ax-b|| 2.22133e-06 ||b|| 0.390066 ||Ax-b||/||b|| 5.69474e-06
norms: 0.012748
1 processors solved in 8.99125 secs with 5 iterations to 2.22133e-06 norm matrix "fidapm11.mtx" vector "fidapm11_rhs1.mtx" true residual ||Ax-b||/||b|| 5.69474e-06
reason: converged due to relative tolerance

$ ./MatSolve 7 fidapm11{,_rhs1}.mtx
load matrix: 0.919449
load vector: 0.0251751
 DLS =    111837342.40956829       DLR =    742304.87677046272     
 AFNORM =    149.31175439327785     
 LFNORM =    8885.2384498644169       UFNORM =    1493.5602536863867     
preconditioner: 5.93604
ILU: iters:   5, residual=0.195E-05 (|SOL|= 0.3141E+03, -0.1525E+01 < SOL <  0.7790E+01)
solver: 0.155
||Ax-b|| 1.94846e-06 ||b|| 0.390066 ||Ax-b||/||b|| 4.9952e-06
norms: 0.013761
1 processors solved in 7.04531 secs with 5 iterations to 1.94846e-06 norm matrix "fidapm11.mtx" vector "fidapm11_rhs1.mtx" true residual ||Ax-b||/||b|| 4.9952e-06
reason: Unspecified for ANI

$ ./MatSolve 0 utm5940{,_rhs1}.mtx
load matrix: 0.118648
load vector: 0.00725484
preconditioner: 0.557577
solver: 1.25405 6.84627e-07 | 0
||Ax-b|| 6.84669e-07 ||b|| 0.0787991 ||Ax-b||/||b|| 8.6888e-06
norms: 0.00134802
1 processors solved in 1.93836 secs with 242 iterations to 6.84627e-07 norm matrix "utm5940.mtx" vector "utm5940_rhs1.mtx" true residual ||Ax-b||/||b|| 8.6888e-06
reason: converged due to relative tolerance

$ ./MatSolve 7 utm5940{,_rhs1}.mtx
load matrix: 0.127311
load vector: 0.00683498
 DLS =    46366.953650296920       DLR =    21752.013525400613     
 AFNORM =    77.071395471990556     
 LFNORM =    5102548.6446918538       UFNORM =    36863459.727390870     
preconditioner: 0.471755
ILU: iters: 209, residual=0.355E-06 (|SOL|= 0.3644E+06, -0.4066E+04 < SOL <  0.3480E+05)
solver: 0.827171
||Ax-b|| 3.54508e-07 ||b|| 0.0787991 ||Ax-b||/||b|| 4.49888e-06
norms: 0.00119901
1 processors solved in 1.43425 secs with 209 iterations to 3.5452e-07 norm matrix "utm5940.mtx" vector "utm5940_rhs1.mtx" true residual ||Ax-b||/||b|| 4.49888e-06
reason: Unspecified for ANI

$ ./MatSolve 0 e40r2000{,_rhs1}.mtx
load matrix: 0.789258
load vector: 0.022099
preconditioner: 5.31056
solver: 1.44443 6.78607e-06 | 0
||Ax-b|| 6.78607e-06 ||b|| 1.81648 ||Ax-b||/||b|| 3.73584e-06
norms: 0.011626
1 processors solved in 7.57202 secs with 46 iterations to 6.78607e-06 norm matrix "e40r2000.mtx" vector "e40r2000_rhs1.mtx" true residual ||Ax-b||/||b|| 3.73584e-06
reason: converged due to relative tolerance

$ ./MatSolve 7 e40r2000{,_rhs1}.mtx
load matrix: 0.779385
load vector: 0.0211039
 DLS =    8843569.9856354389       DLR =    571941177.37418306     
 AFNORM =    131.45721737508106     
 LFNORM =    247819.34259843259       UFNORM =    8116782.0636874922     
preconditioner: 3.99067
ILU: iters:  34, residual=0.178E-04 (|SOL|= 0.5607E+03, -0.2514E+02 < SOL <  0.7768E+02)
solver: 0.915328
||Ax-b|| 1.77709e-05 ||b|| 1.81648 ||Ax-b||/||b|| 9.78317e-06
norms: 0.0117531
1 processors solved in 5.71366 secs with 34 iterations to 1.77709e-05 norm matrix "e40r2000.mtx" vector "e40r2000_rhs1.mtx" true residual ||Ax-b||/||b|| 9.78317e-06
reason: Unspecified for ANI
```

I think at this point we should decide if we still need to support ANI code.

**Comment by kirill-terekhov** [comment]
1.3x required number of iterations for internal solver is suspicious and cannot be attributed only to difference in numerical precision. Probably I have changed some dropping rules or bcg differs somehow. Will look at it later.
